src/contracts/AbstractARM.sol

```diff
@@ -0,0 +1,686 @@
+    /// The accrued fees are from the last time fees were calculated.
+    function _rawTotalAssets() internal view returns (uint256) {
+        // Get the assets in the ARM and external withdrawal queue
+        uint256 assets = token0.balanceOf(address(this)) + token1.balanceOf(address(this)) + _externalWithdrawQueue();
+
+        // Load the queue metadata from storage into memory
+        uint256 queuedMem = withdrawsQueued;
+        uint256 claimedMem = withdrawsClaimed;
+
+        // If the ARM becomes insolvent enough that the total value in the ARM and external withdrawal queue
+        // is less than the outstanding withdrawals.
+        if (assets + claimedMem < queuedMem) {
+            return 0;
+        }
+
+        // Need to remove the liquidity assets that have been reserved for the withdrawal queue
+        // and any accrued fees
+        return assets + claimedMem - queuedMem - feesAccrued;
+    }
+
+    /// @dev Hook for calculating the amount of assets in an external withdrawal queue like Lido or OETH
+    /// This is not the ARM's withdrawal queue
+    function _externalWithdrawQueue() internal view virtual returns (uint256 assets);
+
+    /// @notice Calculates the amount of shares for a given amount of liquidity assets
+    function convertToShares(uint256 assets) public view returns (uint256 shares) {
+        uint256 totalAssetsMem = totalAssets();
+        shares = (totalAssetsMem == 0) ? assets : (assets * totalSupply()) / totalAssetsMem;
+    }
+
+    /// @notice Calculates the amount of liquidity assets for a given amount of shares
+    function convertToAssets(uint256 shares) public view returns (uint256 assets) {
+        assets = (shares * totalAssets()) / totalSupply();
+    }
+
+    /// @notice Set the Liquidity Provider Controller contract address.
+    /// Set to a zero address to disable the controller.
+    function setLiquidityProviderController(address _liquidityProviderController) external onlyOwner {
+        liquidityProviderController = _liquidityProviderController;
+
+        emit LiquidityProviderControllerUpdated(_liquidityProviderController);
+    }
+
+    ////////////////////////////////////////////////////
+    ///         Performance Fee Functions
+    ////////////////////////////////////////////////////
+
+    /// @dev Accrues the performance fee based on the increase in total assets
+    /// Needs to be called before any action that changes the liquidity provider shares. eg deposit and redeem
+    function _accruePerformanceFee() internal {
+        uint256 newTotalAssets = _rawTotalAssets();
+
+        // Do not accrued a performance fee if the total assets has decreased
+        if (newTotalAssets <= lastTotalAssets) return;
+
+        uint256 assetIncrease = newTotalAssets - lastTotalAssets;
+        uint256 newFeesAccrued = (assetIncrease * fee) / FEE_SCALE;
+
+        // Save the new accrued fees back to storage
+        feesAccrued = SafeCast.toUint112(feesAccrued + newFeesAccrued);
```

Comment:
The three storage variables are packed together into one slot for gas savings reason.

However, they are written separately here, which means that we'll still pay two write costs, vs one. We'd probably need to convert it to a struct to get the full benefits.

Reply:
I tried putting the fee data into a struct as follows
```Solidity
    struct FeeData {
        uint16 fee;
        uint112 feesAccrued;
        uint128 lastTotalAssets;
    }

    FeeData public feeData;
```

Most of the LP functions has increased gas costs 

Gas costs before

| Function Name                                                       | min             | avg    | median | max    | # calls |
|--|--|--|--|--|--|
| claimRedeem                                                         | 35800           | 52843  | 52186  | 64407  | 10      |
| collectFees                                                         | 84487           | 102785 | 106035 | 114585 | 4       |
| deposit                                                             | 135407          | 163793 | 164112 | 173747 | 39      |
| requestRedeem                                                       | 160218          | 192799 | 194418 | 199566 | 21      |


Gas costs after
| Function Name                                                       | min             | avg    | median | max    | # calls |
|--|--|--|--|--|--|
| claimRedeem                                                         | 35867           | 52910  | 52253  | 64474   | 10      |
| collectFees                                                         | 82194           | 101983 | 105730 | 114280  | 4       |
| deposit                                                             | 135628          | 163876 | 164453 | 171667  | 39      |
| requestRedeem                                                       | 160452          | 192687 | 194652 | 199800  | 21      |

The gas costs are relatively low compared to other protocols so I'd be included to leave the functions as is at the moment.

Reply:
Agree that this is not worth messing with. Only affects LP functions, not trades.